lib/content_location_filter.js

```diff
@@ -3,15 +3,51 @@
 const HyperSwitch = require('hyperswitch');
 const URI = HyperSwitch.URI;
 const mwUtil = require('./mwUtil');
+const path = require('path');
+const P = require('bluebird');
+
+function isAbsoluteRedirect(location) {
+    return /^https?:/.test(location);
+}
+
+function resolveRelativeRedirect(hyper, req, res) {
+    return mwUtil.getSiteInfo(hyper, req)
+    .then((siteInfo) => {
+        const pathArr = req.uri.path.slice(2).map(encodeURIComponent);
+        pathArr.push('..');
+        pathArr.push(res.headers.location);
+        return `${siteInfo.baseUri}/${path.join.apply(path, pathArr)}`;
+    });
+}
+
+function resolvedRedirectResponse(hyper, req, res) {
+    let getContentURI;
+    if (isAbsoluteRedirect(res.headers.location)) {
+        getContentURI = P.resolve(res.headers.location);
+    } else {
+        getContentURI = resolveRelativeRedirect(hyper, req, res);
+    }
+    return getContentURI.then((contentURI) => hyper.request({
+        method: req.method,
+        uri: new URI(contentURI)
```

Comment:
I *think* this implicitly avoids cyclical redirects by not forwarding the `origin` header, and thus keeping those redirects HTTP-only & iterative. Since this is quite subtle & important, lets add a comment to make sure that we don't break this in the future.